jmclient/test/test_wallet_rpc.py

```diff
@@ -712,6 +718,88 @@ def process_get_seed_response(self, response, code):
         json_body = json.loads(response.decode('utf-8'))
         assert json_body["seedphrase"]
 
+class TrialTestWRPC_JWT(WalletRPCTestBase, unittest.TestCase):
+    @defer.inlineCallbacks
+    def test_refresh_token(self):
+        """Test refresh token endpoint
+
+        1. Valid access, valid refresh -> Success
+        2. Valid access, invalid refresh -> Failed
+        3. Valid access, expired refresh -> Failed
+        4. Expired access, valid refresh -> Success
+        5. Expired access, invalid refresh -> Failed
+        6. Expired access, expired refresh -> Failed
+        7. Invalid access, valid refresh -> Failed
+        """
+
+        agent = get_nontor_agent()
+        addr = self.get_route_root()
+        addr += "/token/refresh"
+        addr = addr.encode()
+
+        for access_token, refresh_token, is_successful_response in [
+            ("valid", "valid", True),
+            ("valid", "invalid", False),
+            ("valid", "expired", False),
+            ("expired", "valid", True),
+            ("expired", "invalid", False),
+            ("expired", "expired", False),
+            ("invalid", "valid", False),
+        ]:
+            access_token, refresh_token, is_successful_response = "valid", "valid", True
+            body = BytesProducer(
+                json.dumps(
+                    {
+                        "refresh_token": self.get_token(
+                            expired=refresh_token == "expired",
+                            invalid=refresh_token == "invalid",
+                            refresh=True,
+                        )
+                    }
+                ).encode()
+            )
+            response_handler = (
+                self.successful_refresh_response_handler
+                if is_successful_response
+                else self.failed_refresh_response_handler
+            )
+            yield self.do_request(
+                agent,
+                b"POST",
+                addr,
+                body,
+                response_handler,
+                self.get_token(
+                    expired=access_token == "expired",
+                    invalid=access_token == "invalid",
+                ),
+            )
+
+    def failed_refresh_response_handler(self, response, code):
+        assert code == 401
+
+    def successful_refresh_response_handler(self, response, code):
+        assert code == 200
+        json_body = json.loads(response.decode("utf-8"))
+        assert json_body.get("token")
+        assert json_body.get("refresh_token")
+
+    def get_token(self, *, expired=False, invalid=False, refresh=False):
+        if invalid:
+            return jwt.encode({"scope": "invalid"}, "12345678")
```

Comment:
@roshii Arguably a nit, but: to test that we don't accept invalid tokens is .. hard; it's not sufficient to give one obviously invalid string.

First, it'd be better to have a whole dataset of invalid token encodings, but that's asking a bit much. If we're only going to test 1 or a few invalid strings, I tend in the opposite direction: create a valid string and malleate one byte (or char).

Reply:
Oh, and I forgot to also mention, you should definitely use a valid *scope* here, right?

Reply:
> Oh, and I forgot to also mention, you should definitely use a valid _scope_ here, right?

Right, valid _scope_ with invalid signature (and opposite) should be tested as well. I will use the proposed approach to create a few more test cases.